generator/regions/regions_builder.hpp

```diff
@@ -1,49 +1,79 @@
+
 #include <functional>
-#include <map>
 #include <memory>
 #include <string>
+#include <utility>
 #include <vector>
 
 namespace generator
 {
 namespace regions
 {
+
+class CountrySpecifier;
+
 // This class is needed to build a hierarchy of regions. We can have several nodes for a region
 // with the same name, represented by a multi-polygon (several polygons).
 class RegionsBuilder
 {
 public:
   using Regions = std::vector<Region>;
+  using RegionPlaceLot = std::vector<RegionPlace>;
   using StringsList = std::vector<std::string>;
-  using IdStringList = std::vector<std::pair<base::GeoObjectId, std::string>>;
-  using CountryTrees = std::multimap<std::string, Node::Ptr>;
-  using NormalizedCountryFn = std::function<void(std::string const &, Node::Ptr const &)>;
+  using CountryFn = std::function<void(std::string const &, Node::PtrList const &,
+                                       CountryRegionsBuilderStats const &)>;
 
-  explicit RegionsBuilder(Regions && regions,
-                          std::unique_ptr<ToStringPolicyInterface> toStringPolicy,
-                          size_t threadsCount = 1);
-  explicit RegionsBuilder(Regions && regions, size_t threadsCount = 1);
+  RegionsBuilder(Regions && regions, PlacePointsMap && placePointsMap, size_t threadsCount = 1);
 
-  Regions const & GetCountries() const;
+  RegionPlaceLot const & GetCountriesOuters() const;
   StringsList GetCountryNames() const;
-  void ForEachNormalizedCountry(NormalizedCountryFn fn);
-  IdStringList ToIdStringList(Node::Ptr const & tree) const;
+  void ForEachCountry(CountryFn const & fn);
 private:
-  static Node::PtrList MakeSelectedRegionsByCountry(Region const & country,
-                                                    Regions const & allRegions);
-  static Node::Ptr BuildCountryRegionTree(Region const & country, Regions const & allRegions);
-  std::vector<Node::Ptr> BuildCountryRegionTrees(RegionsBuilder::Regions const & countries);
-
-  std::unique_ptr<ToStringPolicyInterface> m_toStringPolicy;
-  Regions m_countries;
-  Regions m_regions;
+  using ParentChildPairs = std::vector<std::pair<Node::Ptr, Node::Ptr>>;
+
+  RegionPlaceLot FormRegionPlaceOrder(Regions && regions, PlacePointsMap const & placePointsMap);
+  void EraseLabelPlacePoints(PlacePointsMap & placePointsMap, RegionPlaceLot const & placeOrder);
+  RegionPlaceLot ExtractCountriesOuters(RegionPlaceLot & placeOrder);
+
+  std::vector<Node::Ptr> BuildCountryRegionTrees(RegionPlaceLot const & countryOuters,
+                                                 CountrySpecifier const & countrySpecifier);
+  std::vector<Node::Ptr> MakeCountryNodeOrder(RegionPlace const & countryOuter,
+                                              CountrySpecifier const & countrySpecifier);
+  static ParentChildPairs FindAllParentChildPairs(std::vector<Node::Ptr> const & nodeOrder,
+                                                  std::size_t startOffset, std::size_t step,
```

Comment:
We usually don't write std:: for size_t.

Reply:
done